File: internal/hooks/dispatcher.go
package hooks

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math"
	"net/http"
	"sync"
	"time"
)

const (
	// DefaultMaxRetries is the default number of delivery retry attempts.
	DefaultMaxRetries = 3
	// DefaultBaseDelay is the base delay for exponential backoff.
	DefaultBaseDelay = 1 * time.Second
	// DefaultMaxDelay caps the exponential backoff delay.
	DefaultMaxDelay = 30 * time.Second
	// DefaultHTTPTimeout is the timeout for HTTP delivery requests.
	DefaultHTTPTimeout = 10 * time.Second
	// DefaultWorkerCount is the number of concurrent delivery workers.
	DefaultWorkerCount = 4
)

// DeliveryCallback is called after each delivery attempt (success or failure).
type DeliveryCallback func(d *Delivery)

// Dispatcher handles async webhook delivery with exponential backoff retries.
type Dispatcher struct {
	client     *http.Client
	maxRetries int
	baseDelay  time.Duration
	maxDelay   time.Duration
	workers    int
	wg         sync.WaitGroup
	stopCh     chan struct{}
}

// DispatcherOption configures the Dispatcher.
type DispatcherOption func(*Dispatcher)

// WithMaxRetries sets the maximum number of retry attempts per delivery.
func WithMaxRetries(n int) DispatcherOption {
	return func(d *Dispatcher) { d.maxRetries = n }
}

// WithBaseDelay sets the base delay for exponential backoff.
func WithBaseDelay(delay time.Duration) DispatcherOption {
	return func(d *Dispatcher) { d.baseDelay = delay }
}

// WithHTTPTimeout sets the HTTP client timeout.
func WithHTTPTimeout(timeout time.Duration) DispatcherOption {
	return func(d *Dispatcher) { d.client.Timeout = timeout }
}

// WithWorkerCount sets the number of concurrent delivery goroutines.
func WithWorkerCount(n int) DispatcherOption {
	return func(d *Dispatcher) { d.workers = n }
}

// NewDispatcher creates a new delivery dispatcher with the given options.
func NewDispatcher(opts ...DispatcherOption) *Dispatcher {
	d := &Dispatcher{
		client:     &http.Client{Timeout: DefaultHTTPTimeout},
		maxRetries: DefaultMaxRetries,
		baseDelay:  DefaultBaseDelay,
		maxDelay:   DefaultMaxDelay,
		workers:    DefaultWorkerCount,
		stopCh:     make(chan struct{}),
	}
	for _, opt := range opts {
		opt(d)
	}
	return d
}

// Start launches the background delivery worker pool.
func (d *Dispatcher) Start() {
	// Workers are started on-demand in Deliver; Start just marks as ready.
	log.Printf("hooks dispatcher started: workers=%d max_retries=%d", d.workers, d.maxRetries)
}

// Stop signals workers to stop and waits for in-flight deliveries.
func (d *Dispatcher) Stop() {
	close(d.stopCh)
	d.wg.Wait()
	log.Printf("hooks dispatcher stopped")
}

// Deliver enqueues a payload for async delivery to the given hook.
// It returns immediately; delivery runs in a background goroutine.
func (d *Dispatcher) Deliver(hook Hook, payload Payload, callback DeliveryCallback) {
	d.wg.Add(1)
	go func() {
		defer d.wg.Done()
		d.deliverWithRetry(hook, payload, callback)
	}()
}

// deliverWithRetry sends the payload with exponential backoff retries.
func (d *Dispatcher) deliverWithRetry(hook Hook, payload Payload, callback DeliveryCallback) {
	body, err := json.Marshal(payload)
	if err != nil {
		log.Printf("hooks: failed to marshal payload for hook %s: %v", hook.ID, err)
		d.recordAttempt(hook, payload.EventType, DeliveryFailed, 0, 0, err.Error(), callback)
		return
	}

	deliveryID := generateID("deliv")

	for attempt := 1; attempt <= d.maxRetries+1; attempt++ {
		// Check for shutdown
		select {
		case <-d.stopCh:
			d.recordAttempt(hook, payload.EventType, DeliveryPending, attempt, d.maxRetries+1, "dispatcher stopped", callback)
			return
		default:
		}

		statusCode, err := d.sendRequest(hook, body)

		if err == nil && statusCode >= 200 && statusCode < 300 {
			// Success
			d.recordAttemptWithID(deliveryID, hook, payload.EventType, DeliverySuccess, statusCode, attempt, d.maxRetries+1, "", callback)
			return
		}

		// Determine error message
		errMsg := fmt.Sprintf("status %d", statusCode)
		if err != nil {
			errMsg = err.Error()
		}

		// If this was the last attempt, record failure
		if attempt > d.maxRetries {
			d.recordAttemptWithID(deliveryID, hook, payload.EventType, DeliveryFailed, statusCode, attempt, d.maxRetries+1, errMsg, callback)
			return
		}

		// Exponential backoff before retry
		delay := d.computeBackoff(attempt)
		d.recordAttemptWithID(deliveryID, hook, payload.EventType, DeliveryRetry, statusCode, attempt, d.maxRetries+1, errMsg, callback)

		select {
		case <-time.After(delay):
		case <-d.stopCh:
			d.recordAttemptWithID(deliveryID, hook, payload.EventType, DeliveryPending, statusCode, attempt, d.maxRetries+1, "dispatcher stopped during retry", callback)
			return
		}
	}
}

// sendRequest performs the actual HTTP POST to the webhook endpoint.
// Returns the status code and any error.
func (d *Dispatcher) sendRequest(hook Hook, body []byte) (int, error) {
	req, err := http.NewRequest(http.MethodPost, hook.URL, bytes.NewReader(body))
	if err != nil {
		return 0, fmt.Errorf("create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "hermes-agent-cluster/1.0")
	req.Header.Set("X-Hook-ID", hook.ID)

	// HMAC-SHA256 signature if secret is configured
	if hook.Secret != "" {
		signature := SignPayload(body, hook.Secret)
		req.Header.Set("X-Hub-Signature-256", signature)
	}

	resp, err := d.client.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	// Drain the body so the connection can be reused
	io.Copy(io.Discard, resp.Body)

	return resp.StatusCode, nil
}

// computeBackoff calculates the delay for the given attempt (1-indexed).
// Uses exponential backoff with jitter, capped at maxDelay.
func (d *Dispatcher) computeBackoff(attempt int) time.Duration {
	// 2^(attempt-1) * baseDelay
	delay := float64(d.baseDelay) * math.Pow(2, float64(attempt-1))
	if delay > float64(d.maxDelay) {
		delay = float64(d.maxDelay)
	}
	return time.Duration(delay)
}

// recordAttempt is a convenience wrapper around recordAttemptWithID.
func (d *Dispatcher) recordAttempt(hook Hook, eventType EventType, status DeliveryStatus, attempts, maxAttempts int, errMsg string, callback DeliveryCallback) {
	d.recordAttemptWithID(generateID("deliv"), hook, eventType, status, 0, attempts, maxAttempts, errMsg, callback)
}

// recordAttemptWithID creates a Delivery record and invokes the callback.
func (d *Dispatcher) recordAttemptWithID(id string, hook Hook, eventType EventType, status DeliveryStatus, statusCode, attempts, maxAttempts int, errMsg string, callback DeliveryCallback) {
	now := time.Now()
	delivery := &Delivery{
		ID:          id,
		HookID:      hook.ID,
		EventType:   eventType,
		URL:         hook.URL,
		Status:      status,
		StatusCode:  statusCode,
		Error:       errMsg,
		Attempts:    attempts,
		MaxAttempts: maxAttempts,
		CreatedAt:   now,
		UpdatedAt:   now,
	}
	if callback != nil {
		callback(delivery)
	}
}

File: hermes_cluster/hooks/dispatcher.py
"""Dispatcher — async webhook delivery with exponential backoff retries (from hooks/dispatcher.go)."""

from __future__ import annotations

import asyncio
import json
import logging
import math
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Optional
from urllib.request import Request, urlopen
from urllib.error import URLError, HTTPError

from hermes_cluster.hooks.payload import (
    DeliveryStatus,
    HookEventType,
    Payload,
    sign_payload,
)

logger = logging.getLogger(__name__)

# Defaults from Go dispatcher.go
DEFAULT_MAX_RETRIES = 3
DEFAULT_BASE_DELAY = 1.0  # seconds
DEFAULT_MAX_DELAY = 30.0  # seconds
DEFAULT_HTTP_TIMEOUT = 10  # seconds
DEFAULT_WORKER_COUNT = 4


@dataclass
class DeliveryRecord:
    """A record of a single delivery attempt."""
    id: str
    hook_id: str
    event_type: str
    url: str
    status: str  # DeliveryStatus value
    status_code: int = 0
    error: str = ""
    attempts: int = 0
    max_attempts: int = 0
    created_at: float = field(default_factory=time.time)
    updated_at: float = field(default_factory=time.time)


# Type for delivery callbacks
DeliveryCallback = Callable[[DeliveryRecord], None]


class Dispatcher:
    """Handles async webhook delivery with exponential backoff retries.

    Mirrors Go's dispatcher.go implementation:
    - Worker pool for concurrent deliveries
    - Exponential backoff with jitter on failure
    - HMAC-SHA256 signing if secret configured
    - Delivery callback for recording results
    """

    def __init__(
        self,
        max_retries: int = DEFAULT_MAX_RETRIES,
        base_delay: float = DEFAULT_BASE_DELAY,
        max_delay: float = DEFAULT_MAX_DELAY,
        http_timeout: int = DEFAULT_HTTP_TIMEOUT,
        worker_count: int = DEFAULT_WORKER_COUNT,
    ):
        self.max_retries = max_retries
        self.base_delay = base_delay
        self.max_delay = max_delay
        self.http_timeout = http_timeout
        self.worker_count = worker_count
        self._running = False
        self._semaphore: Optional[asyncio.Semaphore] = None
        self._tasks: list[asyncio.Task] = []

    def start(self) -> None:
        """Mark dispatcher as ready. Workers are launched on-demand."""
        self._running = True
        self._semaphore = asyncio.Semaphore(self.worker_count)
        logger.info(
            "hooks dispatcher started: workers=%d max_retries=%d",
            self.worker_count,
            self.max_retries,
        )

    def stop(self) -> None:
        """Signal workers to stop."""
        self._running = False
        logger.info("hooks dispatcher stopped")

    async def deliver(
        self,
        hook_id: str,
        hook_url: str,
        hook_secret: str,
        payload: Payload,
        callback: Optional[DeliveryCallback] = None,
    ) -> DeliveryRecord:
        """Deliver payload to a webhook endpoint with retries.

        Returns the final DeliveryRecord.
        """
        body = json.dumps(payload.model_dump(), default=str).encode("utf-8")
        delivery_id = _generate_id("deliv")
        max_attempts = self.max_retries + 1

        for attempt in range(1, max_attempts + 1):
            if not self._running:
                record = DeliveryRecord(
                    id=delivery_id,
                    hook_id=hook_id,
                    event_type=payload.event_type.value,
                    url=hook_url,
                    status=DeliveryStatus.PENDING.value,
                    attempts=attempt,
                    max_attempts=max_attempts,
                    error="dispatcher stopped",
                )
                if callback:
                    callback(record)
                return record

            status_code, error = await self._send_request(
                hook_url, hook_secret, body
            )

            err_msg = f"status {status_code}" if error is None else str(error)

            if error is None and 200 <= status_code < 300:
                record = DeliveryRecord(
                    id=delivery_id,
                    hook_id=hook_id,
                    event_type=payload.event_type.value,
                    url=hook_url,
                    status=DeliveryStatus.SUCCESS.value,
                    status_code=status_code,
                    attempts=attempt,
                    max_attempts=max_attempts,
                )
                if callback:
                    callback(record)
                return record

            # If this was the last attempt, record failure
            if attempt >= max_attempts:
                record = DeliveryRecord(
                    id=delivery_id,
                    hook_id=hook_id,
                    event_type=payload.event_type.value,
                    url=hook_url,
                    status=DeliveryStatus.FAILED.value,
                    status_code=status_code,
                    error=err_msg,
                    attempts=attempt,
                    max_attempts=max_attempts,
                )
                if callback:
                    callback(record)
                return record

            # Record retry attempt
            if callback:
                callback(DeliveryRecord(
                    id=delivery_id,
                    hook_id=hook_id,
                    event_type=payload.event_type.value,
                    url=hook_url,
                    status=DeliveryStatus.RETRY.value,
                    status_code=status_code,
                    error=err_msg,
                    attempts=attempt,
                    max_attempts=max_attempts,
                ))

            # Exponential backoff before next attempt
            delay = self._compute_backoff(attempt)
            await asyncio.sleep(delay)

        # Should not reach here, but safety fallback
        record = DeliveryRecord(
            id=delivery_id,
            hook_id=hook_id,
            event_type=payload.event_type.value,
            url=hook_url,
            status=DeliveryStatus.FAILED.value,
            error="exhausted retries",
            attempts=max_attempts,
            max_attempts=max_attempts,
        )
        if callback:
            callback(record)
        return record

    async def _send_request(
        self, url: str, secret: str, body: bytes
    ) -> tuple[int, Optional[str]]:
        """Perform HTTP POST to webhook endpoint.

        Returns (status_code, error_message).
        """
        headers = {
            "Content-Type": "application/json",
            "User-Agent": "hermes-agent-cluster/1.0",
        }

        # HMAC-SHA256 signature if secret configured
        if secret:
            signature = sign_payload(body, secret)
            headers["X-Hub-Signature-256"] = signature

        try:
            req = Request(url, data=body, headers=headers, method="POST")
            with urlopen(req, timeout=self.http_timeout) as resp:
                return resp.status, None
        except HTTPError as e:
            return e.code, str(e)
        except URLError as e:
            return 0, str(e)
        except Exception as e:
            return 0, str(e)

    def _compute_backoff(self, attempt: int) -> float:
        """Calculate delay with exponential backoff, capped at max_delay."""
        delay = self.base_delay * math.pow(2, attempt - 1)
        return min(delay, self.max_delay)


def _generate_id(prefix: str) -> str:
    """Generate a random hex ID."""
    import secrets
    return f"{prefix}_{secrets.token_hex(8)}"
